### PR TITLE
Run the sharing optimisation in beam_jump until fixpoint

### PR DESCRIPTION
**Pull request by michalmuskala**
This is especially useful after inlining a function with a case.

Today the compiler would most probably be able to unify all the leafs of the
case during the sharing optimisation, but it would fail to unify the pattern
matching itself.

Naively running the optimisation multiple times wouldn't be able to find the
common code either, because it would differ in jump/fail targets of various
instructions.

To remedy this, after doing each sharing pass we traverse the code backwards
when reversing and update all the jump targets with the new targets that were
discovered during the unification pass. This allows running the optimisation
until fixpoint and makes sure all sharing opportunities will be discovered.

This optimisation also helps with the Elixir's `with/else` construct.

Introduce `beam_utils:replace_labels/4` and use in `beam_jump` & `beam_clean`

---

I run the compiler & stdlib test suites without failures.

### CODE REVIEW

**Comment by michalmuskala** [comment]
Playing a bit more with this, I noticed that in many cases the optimisation is prevented by the `kill`/`trim` instructions that differ when `beam_jump` is executed but are the same after running through `beam_trim`. I'm not sure if/how this could be improved.

**Comment by bjorng** [comment]
This looks promising.

However, I was confused how the commits are organized. The first (bottom-most) commit adds `beam_jump:replace_labels/4` and then the third (top-most) commit removes that function again.

If you feel comfortable rearranging code between commits, I would much prefer if the commits were arranged like this:

* One or more commits at the bottom that introduce `beam_utils:replace_labels/4` and use it in `beam_clean`.

* On top, the commit that does the new fixpoint iteration of the sharing optimization.

If that is too complicate/too much work, I would prefer that you squash the commits.

**Comment by michalmuskala** [comment]
@bjorng I restructured it to be in two commits - one that extracts `replace_labels` from `beam_clean` to `beam_utils` and another one that implements the sharing optimisation until fixpoint thing.

File: lib/compiler/src/beam_jump.erl [hunk 1]
```diff
@@ -156,49 +156,62 @@ function({function,Name,Arity,CLabel,Asm0}) ->
...
 	false ->
-	    share_1(Is, Dict, [I|Seq], Acc);
+	    share_1(Is, Dict, Lbls, [I|Seq], Acc);
 	true ->
-	    share_1(Is, Dict, [I], Acc)
+	    share_1(Is, Dict, Lbls, [I], Acc)
     end.
 
-clean_non_sharable(Dict) ->
+clean_non_sharable(Dict0, Lbls0) ->
```
Comment (bjorng) [inline lib/compiler/src/beam_jump.erl]: Can you explain what you are doing with `Lbls` in this function? I am confused about why you delete the label if the   sequence is sharable with `try` and not the other way round.

**Comment by bjorng** [comment]
It seems that in most cases the code is improved. However, I found a few counter examples where the code got worse:

```
diff -u old/code_server.S new/code_server.S
--- old/code_server.S	2017-08-10 05:41:08.000000000 +0200
+++ new/code_server.S	2017-08-10 05:40:35.000000000 +0200
@@ -2785,7 +2785,9 @@
     {test,is_atom,{f,14},[{x,0}]}.
     {select_val,{x,0},{f,14},{list,[{atom,true},{f,11},{atom,false},{f,10}]}}.
   {label,9}.
-    {test,is_tagged_tuple,{f,10},[{x,0},2,{atom,'EXIT'}]}.
+    {test,is_tuple,{f,10},[{x,0}]}.
+    {test,test_arity,{f,10},[{x,0},2]}.
+    {get_tuple_element,{x,0},0,{x,1}}.
   {label,10}.
     {move,{literal,{error,bad_directory}},{x,0}}.
     {deallocate,8}.

diff -u old/dist_ac.S new/dist_ac.S
--- old/dist_ac.S	2017-08-10 05:41:08.000000000 +0200
+++ new/dist_ac.S	2017-08-10 05:40:35.000000000 +0200
@@ -1593,6 +1593,7 @@
     {test,is_tuple,{f,33},[{x,1}]}.
     {test,test_arity,{f,34},[{x,1},2]}.
     {get_tuple_element,{x,1},0,{x,0}}.
+    {test,is_eq_exact,{f,34},[{x,0},{atom,distributed}]}.
     {jump,{f,34}}.
   {label,33}.
     {test,is_eq_exact,{f,34},[{x,1},{atom,local}]}.
```

Perhaps you would like to find out why the code gets worse and see if you can do something about it. I have a guess that it might be because of multiple adjacent labels. Coalescing adjacent labels like it is done in `beam_clean` might help.

Note: I used this script to produce `.S` files that are more diff-friendly:

https://gist.github.com/bjorng/cfb0125cf93375e616231adf5531855d#file-otp-diffable-asm

**Comment by michalmuskala** [comment]
@bjorng I tried running the script on master & the PR but I didn't see any changes. I assume this is because I'm not running the tool correctly. 

When I run the script, it creates a lot of `.S` files in the top-level dir and then in the "target" directory. What files should I use to compare?

**Comment by bjorng** [comment]
The ones in the target directory.

Did you make sure to the set the PATH?

Here is what I typically do:
```
PATH=$ERL_TOP/bin:$PATH otp-diffable-asm new
git checkout HEAD~2
make
PATH=$ERL_TOP/bin:$PATH otp-diffable-asm old
diff -u old new

```

(Note: I use `git checkout HEAD~2` because there are two commits in your branch.)

**Comment by michalmuskala** [comment]
I don't think this is caused by multiple labels - at the point, this optimisation is applied, we don't have multiple neighbouring labels. This is introduced only in later passes of `beam_jump`.

I found something else, though. There are other places where something similar to what you highlighted in `code_server` is happening, e.g.:

```diff
diff -u old/test_server.S new/test_server.S
--- old/test_server.S   2017-08-12 18:45:50.000000000 +0200
+++ new/test_server.S   2017-08-12 18:50:46.000000000 +0200
@@ -3383,7 +3383,9 @@
     {kill,{y,3}}.
     {kill,{y,4}}.
     {call,4,{f,{do_end_tc_call,4}}}.
-    {test,is_tagged_tuple,{f,4},[{x,0},2,{atom,failed}]}.
+    {test,is_tuple,{f,4},[{x,0}]}.
+    {test,test_arity,{f,4},[{x,0},2]}.
+    {get_tuple_element,{x,0},0,{x,1}}.
   {label,4}.
     {test_heap,3,1}.
     {put_tuple,2,{x,1}}.
```

Interestingly, there's also another change that pops up frequently and I think is related:
```diff
diff -u old/typer.S new/typer.S
--- old/typer.S 2017-08-12 18:45:54.000000000 +0200
+++ new/typer.S 2017-08-12 18:50:49.000000000 +0200
@@ -5,7 +5,7 @@

 {attributes, []}.

-{labels, 507}.
+{labels, 504}.


 {function, start, 0, {start,0}}.
@@ -1331,22 +1331,20 @@
     {allocate,1,2}.
     {move,{x,0},{y,0}}.
     {call,2,{f,{map__lookup,2}}}.
-    {test,is_tuple,{f,3},[{x,0}]}.
-    {test,test_arity,{f,4},[{x,0},2]}.
+    {test,is_tuple,{f,2},[{x,0}]}.
+    {test,test_arity,{f,3},[{x,0},2]}.
     {get_tuple_element,{x,0},0,{x,1}}.
-    {test,is_eq_exact,{f,2},[{x,1},{atom,contract}]}.
-  {label,2}.
     {deallocate,1}.
     return.
-  {label,3}.
-    {test,is_eq_exact,{f,4},[{x,0},{atom,none}]}.
+  {label,2}.
+    {test,is_eq_exact,{f,3},[{x,0},{atom,none}]}.
     {test_heap,2,0}.
     {put_list,{y,0},nil,{x,1}}.
     {trim,1,0}.
     {move,{literal,"No type info for function: ~tp\n"},{x,0}}.
     {call_ext,2,{extfunc,io_lib,format,2}}.
     {call_last,1,{f,{fatal_error,1}},0}.
-  {label,4}.
+  {label,3}.
     {case_end,{x,0}}.
```

It looks as if the compiler eliminated the check for the tuple tag. I think this is also the reason we loose the `is_tagged_tuple` optimisation - in all the places where it's lost, we no longer have the check for the value of the tag.

I will try to figure out why that is happening.

**Comment by michalmuskala** [comment]
I also had another idea how to make the sharing optimisation even better. In a lot of cases, the only difference between blocks of code is some constant value - most of the time an atom or a string. My idea was to somehow yank the constants out of the fragments before the unification. This, in theory, would make it possible to unify fragments that only differ by some constant. If the unification is successful we can reintroduce the constants using `move` instructions and jumps into the common fragments.

**Comment by michalmuskala** [comment]
I understand why sequence in `code_server` ends up like this. This condition is forced by the following code:
```erlang
case catch lists:reverse(filename:split(Dir)) of
    {'EXIT', _} ->
        {error, bad_directory};
    %% other clauses, none matches on a tuple
    _ ->
        {error, bad_directory}
end
```

After the unification pass, both error branches are unified and the asm around the `'EXIT'` check looks, more-or-less like this:

```S
  {test,is_eq_exact,{f,19},[{x,1},{atom,'EXIT'}]}
  {jump,{f,19}},
{label,40},
  {jump,{f,19}},
{label,19},
  {block,[{set,[{x,0}],[{literal,{error,bad_directory}}],move}]},
  {deallocate,8},
```

This means we hit the following optimisation: https://github.com/erlang/otp/blob/master/lib/compiler/src/beam_jump.erl#L295-L309

The compiler notices, the test is useless and eliminates it.

The proper fix, in my opinion, would be to see if after removing the tests in `beam_jump` any of the registers removed in the test is actually used later on. If not we could eliminate the `get_tuple_element` instruction and all the other tests since it actually doesn't matter if it's a tuple or not.

I'm not sure what to do with this. I have a feeling that maybe a better place for eliminating patterns that have exactly the same body would be actually somewhere in the core passes, so that this doesn't actually reach the asm passes - as far as I understand the various asm passes, verifying the register is not used anywhere, so that we can eliminate the `get_tuple_element` would be rather hard in this place.

File: lib/compiler/src/beam_jump.erl [hunk 2]
```diff
@@ -383,6 +399,47 @@ insert_fc_labels([{label,L}=I|Is0], Mlbl) ->
...
+            {[I|Is],St};
+        true ->
+            opt_useless_loads(Is, Args ++ Args1, L, St)
+    end;
+opt_useless_loads(Is, _Args, _L, St) ->
+    {Is,St}.
+
+opt_useless_block_loads([{set,[Dst],_,_}=I|Is], Args, L, Index) ->
+    case lists:member(Dst, Args) andalso
+        (Is =:= [] orelse beam_utils:is_killed_block(Dst, Is)) andalso
```
Comment (michalmuskala) [inline lib/compiler/src/beam_jump.erl]: @bjorng I'm not sure why `beam_utils:is_killed_block` returns `false` for an empty block.

**Comment by michalmuskala** [comment]
I added another optimisation to `beam_jump` that tries to check if other isntructions could be removed after eliminating a `test` instruction. This fixes the problematic code in `code_server` - now there's no test for a tuple at all.

**Comment by michalmuskala** [comment]
The second issue you highlighted was indeed caused by duplicated labels. I worked around it by replacing the labels in the `opt` pass of `beam_jump` instead of duplicating them.

This also revealed an opportunity to perform the test elimination optimisation also in case we have a fall-through. Even though, there are no fall-throughs when we enter `opt`, it can produce them itself and we're running until fixpoint.

File: lib/compiler/src/beam_jump.erl (hunk 1 above)
Comment (michalmuskala) [inline lib/compiler/src/beam_jump.erl]: To be honest, I don't remember anymore why I did exactly this 😞 

I only know that when I remove this, I get test failures. In particular the "self compile old inliner" test from compiler suite fails with:
```
sys_core_fold: function eval_length/3+99:
  Internal consistency check failed - please report this bug.
  Instruction: {get_tuple_element,{x,0},1,{x,3}}
  Error:       ambiguous_catch_try_state:
```

File: lib/compiler/src/beam_jump.erl (hunk 2 above)
Comment (bjorng) [inline lib/compiler/src/beam_jump.erl]: `false` means that the register is either not killed OR that we don't know.

Comment (bjorng) [inline lib/compiler/src/beam_jump.erl]: Not sure why this entire line is needed.

The `set` instruction already kills `Dst`, so why test that it is killed in the rest of the block?

**Comment by bjorng** [comment]
I have pushed a fixup commit, which simplifies the code. It can also remove a `get_tuple_element` that is not at the end of a block:

```
 diff -u new/beam_dead.S newer/beam_dead.S
--- new/beam_dead.S     2017-08-14 10:40:33.701140873 +0200
+++ newer/beam_dead.S   2017-08-14 11:42:27.457158130 +0200
@@ -2017,7 +2017,6 @@
     {test,is_eq_exact,{f,10},[{x,5},{atom,bs_save2}]}.
     {test,is_tuple,{f,10},[{x,1}]}.
     {test,test_arity,{f,10},[{x,1},2]}.
-    {get_tuple_element,{x,3},1,{x,6}}.
     {get_tuple_element,{x,3},2,{x,7}}.
     {jump,{f,10}}.
   {label,4}.
```

If you like this fix, please squash it. You can also further simplify the code by removing `lists:member(Dst, Args) andalso` and the `Args` argument. (You will have to resolve some merge conflicts in order to keep the commits clean.)

**Comment by bjorng** [comment]
Added to our daily builds.

**Pull request merged**